test/compiler/inference.jl

```diff
@@ -3721,3 +3721,63 @@ end |> only == Tuple{Int,Int}
     s2 = Some{Any}(s1)
     s2.value.value
 end |> only == Int
+
+# issue #42986
+@testset "narrow down `Union` using `isdefined` checks" begin
+    # basic functionality
+    @test Base.return_types((Union{Nothing,Core.CodeInstance},)) do x
+        if isdefined(x, :inferred)
+            return x
+        else
+            throw("invalid")
+        end
+    end |> only === Core.CodeInstance
+
+    @test Base.return_types((Union{Nothing,Core.CodeInstance},)) do x
+        if isdefined(x, :not_exist)
+            return x
+        else
+            throw("invalid")
+        end
+    end |> only === Union{}
+
+    # > 2 types in the union
+    let
+        struct UnionNarrowingByIsdefinedA; x; end
+        struct UnionNarrowingByIsdefinedB; x; end
+        struct UnionNarrowingByIsdefinedC; x; end
+        @test  Base.return_types((Union{UnionNarrowingByIsdefinedA, UnionNarrowingByIsdefinedB, UnionNarrowingByIsdefinedC},)) do x
+            if isdefined(x, :x)
+                return x
+            else
+                throw("invalid")
+            end
+        end |> only === Union{UnionNarrowingByIsdefinedA, UnionNarrowingByIsdefinedB, UnionNarrowingByIsdefinedC}
+    end
+
+    # > 2 types in the union and some aren't defined
+    let
+        struct UnionNarrowingByIsdefinedA; x; end
+        struct UnionNarrowingByIsdefinedC; x; end
+        @test  Base.return_types((Union{UnionNarrowingByIsdefinedA, Core.CodeInstance, UnionNarrowingByIsdefinedC},)) do x
+            if isdefined(x, :x)
+                return x
+            else
+                throw("invalid")
+            end
+        end |> only === Union{UnionNarrowingByIsdefinedA, UnionNarrowingByIsdefinedC}
+    end
+
+    # should respect `Const` information still
+    let
+        struct UnionNarrowingByIsdefinedA; x; end
+        struct UnionNarrowingByIsdefinedB; x; end
```

Comment:
These `struct` definitions are essentially duplicated (in Julia struct definitions are always global even if they appear as local). So let's unify and define them all at the top of `@testset` block.